### DISCUSSION TITLE
Event binding issue with an ItemsControl (MVVM & code behind issues)

### DISCUSSION THREAD

**Discussion by simon-moisy** [AvaloniaUI/Avalonia#7739] [Q&A] [answered]
First of all, I am relatively new to MVVM and Avalonia so I might be asking something obvious. Apologies in advances if this is the case.

I have a view containing an ItemsControl.I bind it's Items property to an ObservableCollection of Animation (custom class). 
I want to be able to detect PointerPressed on any of those displayed Animation. I've started by registering to the PointerPressed event (my ItemsControl.ItemTemplate displays the Animations inside a StackPanel). The problem is that the event handler "sender" is of StackPanel type, I would like to get a hand on the Animation that has been clicked instead. Is that doable ?

Here's my ItemsControl XAML code :

```
<ItemsControl Grid.Row="2" Items="{Binding AnimationSequence}">
    <ItemsControl.ItemsPanel>
        <ItemsPanelTemplate>
            <StackPanel Orientation="Horizontal" HorizontalAlignment="Center" />
        </ItemsPanelTemplate>
    </ItemsControl.ItemsPanel>

    <ItemsControl.ItemTemplate>
        <DataTemplate DataType="{x:Type ui:Animation}">
            <StackPanel PointerPressed="InputElement_OnPointerPressed">
                <Panel>
                    <Rectangle Width="{Binding DisplayWidth}" Fill="{Binding Color}" />
                    <Label VerticalAlignment="Center" Width="{Binding DisplayWidth}">
                        <TextBlock Text="{Binding AnimationName}"/>
                    </Label>
                </Panel>
            </StackPanel>
        </DataTemplate>
    </ItemsControl.ItemTemplate>
</ItemsControl>
```

In a second time, I would like to implement it in MVVM style, and I have no idea how to do so. I've registered to other pointer events in MVVM style by doing the following, but I don't think I can do it this way with my ItemsControl's Items since it's templated :

```
var ctrl = this.FindControl<ItemsControl>("AnimationSequenceControl");

this.WhenActivated(d => {
    var ctrlPointerMoved = Observable.FromEventPattern(ctrl, nameof(ctrl.PointerMoved));
    var ctrlPointerReleased = Observable.FromEventPattern(ctrl, nameof(ctrl.PointerReleased));
    var ctrlPointerLeft = Observable.FromEventPattern(ctrl, nameof(ctrl.PointerLeave));

    ctrlPointerMoved
        .Select(e => Unit.Default)
        .InvokeCommand(this, x => x.ViewModel.DragUpdateCommand);
    ctrlPointerReleased
        .Select(e => true)
        .InvokeCommand(this, x => x.ViewModel.DragEndCommand);
    ctrlPointerLeft
        .Select(e => false)
        .InvokeCommand(this, x => x.ViewModel.DragEndCommand);
});
```

Thanks in advance, any help would be much appreciated !

**Comment by wieslawsoltes** [comment, marked as answer]
You can use behaviors https://github.com/wieslawsoltes/AvaloniaBehaviors

Samples https://github.com/wieslawsoltes/AvaloniaBehaviors/tree/master/samples

**Reply by simon-moisy** [reply]
Thank you, I will take a look at those samples.

**Reply by simon-moisy** [reply]
This is awesome, exactly what I was looking for (in particular the drag and drop feature...)
Thanks again !

**Reply by simon-moisy** [reply]
@wieslawsoltes how would you use the ItemsDragAndDrop on an ItemsControl ? I see you define the style like this
``` <Style Selector="ListBox.ItemsDragAndDrop ListBoxItem">```
but an equivalent of ListBoxItem for an ItemsControl doesn't seem to exist

**Reply by wieslawsoltes** [reply]
Did not try if this would work but `<Style Selector="ItemsControl.ItemsDragAndDrop ContentPresenter">`

**Reply by simon-moisy** [reply]
It does. Thank you, appreciated your help a lot !
